server/src/main/java/org/elasticsearch/action/admin/indices/diskusage/IndexDiskUsageAnalyzer.java

```diff
@@ -0,0 +1,648 @@
+                    }
+                    sorted.lookupOrd(0);
+                    sorted.lookupOrd(sorted.getValueCount() - 1);
+                    break;
+                case SORTED_SET:
+                    final SortedSetDocValues sortedSet = docValuesReader.getSortedSet(fieldInfo);
+                    while (sortedSet.nextDoc() != DocIdSetIterator.NO_MORE_DOCS) {
+                        while (sortedSet.nextOrd() != SortedSetDocValues.NO_MORE_ORDS) {
+                            cancellationChecker.logEvent();
+                        }
+                    }
+                    sortedSet.lookupOrd(0);
+                    sortedSet.lookupOrd(sortedSet.getValueCount() - 1);
+                    break;
+                default:
+                    assert false : "Unknown docValues type [" + dvType + "]";
+                    throw new IllegalStateException("Unknown docValues type [" + dvType + "]");
+            }
+            stats.addDocValues(fieldInfo.name, directory.getBytesRead());
+        }
+    }
+
+    void analyzePostings(SegmentReader reader, IndexDiskUsageStats stats) throws IOException {
+        // TODO: FieldsReader has stats() which might contain the disk usage infos
+        // Also, can we track the byte reader per field extension to avoid visiting terms multiple times?
+        FieldsProducer postingsReader = reader.getPostingsReader();
+        if (postingsReader == null) {
+            return;
+        }
+        postingsReader = postingsReader.getMergeInstance();
+        PostingsEnum postings = null;
+        for (FieldInfo field : reader.getFieldInfos()) {
+            if (field.getIndexOptions() == IndexOptions.NONE) {
+                continue;
+            }
+            cancellationChecker.checkForCancellation();
+            directory.resetBytesRead();
+            final Terms terms = postingsReader.terms(field.name);
+            if (terms == null) {
+                continue;
+            }
+
+            // Visit terms index and term dictionary and postings
+            final BytesRefIterator termsIterator = terms.iterator();
+            TermsEnum termsEnum = terms.iterator();
+            BytesRef bytesRef;
+            while ((bytesRef = termsIterator.next()) != null) {
+                cancellationChecker.logEvent();
+                termsEnum.seekExact(bytesRef); // it's expensive to look up every term
+                termsEnum.totalTermFreq();
+                for (long idx = 0; idx <= 8; idx++) {
+                    cancellationChecker.logEvent();
+                    final int skipDocID = Math.toIntExact(idx * reader.maxDoc() / 8);
+                    postings = termsEnum.postings(postings, PostingsEnum.NONE);
+                    if (postings.advance(skipDocID) != DocIdSetIterator.NO_MORE_DOCS) {
+                        postings.freq();
+                        postings.nextDoc();
+                    } else {
+                        break;
+                    }
```

Comment:
My understanding is that we feel good about not iterating over all docs because we track the min/max positions rather than which bytes exactly are being accessed, could you leave a comment about it if my understanding is correct?

Reply:
see https://github.com/elastic/elasticsearch/pull/74051/commits/0efab4490ed81f5ad8779e3cdd0010d324098b98